lib/puppet/provider/ssh_authorized_key/parsed.rb

```diff
@@ -56,16 +55,21 @@ def flush
     # so calling it here suppresses the later attempt by our superclass's flush method.
     self.class.backup_target(target)
 
-    Puppet::Util::SUIDManager.asuser(@resource.should(:user)) do
-      unless Puppet::FileSystem.exist?(dir = File.dirname(target))
-        Puppet.debug "Creating #{dir} as #{@resource.should(:user)}"
-        Dir.mkdir(dir, dir_perm)
+    # if the target path is owned by the root, create the file as root
+    if target_uid == Process.euid
```

Comment:
Unfortunately I don't think this check is sufficient. The target path might be world writable. Or one of the ancestor directories could be owned by non-root, so could be renamed and a new directory tree created in its place. I honestly don't know of a way to do what you're trying to do securely. Thoughts @puppetlabs/security-review?

Reply:
It seems like the goal is to allow writing a user-specific authorized key in a global location. Puppet has to be running as root already for that to make sense; the permissions on the key have to be user-specific for it to be functional.

It seems like it would only make sense to write to directories that either the target user owns or that are owned by root (and thus we know are somewhat "safe"). So would it be ok to conditionally operate based on that?

Reply:
> that are owned by root (and thus we know are somewhat "safe")

The trouble is that an ancestor directory may not be owned by root. And after checking `target_uid == Process.euid`, the ancestory directory could be renamed and new directories/files/symlinks created in its place.

Reply:
Fair point. So entire ancestor tree would have to be owned by root? Or operate on file descriptors (which may not be very portable).

Reply:
In a typical scenario, an user cannot chown files/directories to root, so having a structure like `/home/testuser/a/b/authorized_keys_file` where `b` is owned by root and `a` is owned by the user is an unlikely possibility.

Checking for the ownership of the entire ancestor tree in case we're on the `root` case sounds good to me.

Reply:
I'll investigate this further next week. Thanks for the suggestions!

Reply:
@joshcooper pushed a commit that checks the path is not a symlink and checks if all ancestors are owned by root.

This causes every path that contains symlinks to be treated as not trusted, and the writing will be done as the target user, not root.